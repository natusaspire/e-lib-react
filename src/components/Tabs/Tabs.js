import React, { PureComponent } from 'react';

const ITEM_CLASS_NAME = 'tabs__item';

const SELECTED_ITEM_CLASS_NAME = `${ITEM_CLASS_NAME}--active`;

class Tabs extends PureComponent {
  state = { selectedItem: this.props.items[0] || null };

  selectItem = item => () => {
    const { onItemSelect } = this.props;

    this.setState({ selectedItem: item }, () => onItemSelect(item));
  };

  renderItem = item => (
    <div
      key={item.id}
      className={this.getItemClassName(item.id)}
      onClick={this.selectItem(item)}
    >
      {item.name}
    </div>
  );

  isItemSelected = id => {
    const { selectedItem } = this.state;

    return !!selectedItem && selectedItem.id === id;
  };

  getItemClassName = id =>
    this.isItemSelected(id)
      ? `${ITEM_CLASS_NAME} ${SELECTED_ITEM_CLASS_NAME}`
      : ITEM_CLASS_NAME;

  render() {
    const { items } = this.props;

    return <div className="tabs">{items.map(this.renderItem)}</div>;
  }
}

export default Tabs;
