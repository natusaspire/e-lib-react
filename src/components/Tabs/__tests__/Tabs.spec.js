import React from 'react';
import { shallow } from 'enzyme';

import Tabs from '@/components/Tabs';

describe('<Tabs />', () => {
  const itemClassName = 'tabs__item';
  const selectedItemClassName = `${itemClassName} tabs__item--active`;

  const props = {
    items: [
      {
        id: 1,
        name: 'Text',
        value: 'text'
      },
      {
        id: 2,
        name: 'Audio',
        value: 'audio'
      },
      {
        id: 3,
        name: 'Video',
        value: 'video'
      }
    ],
    onItemSelect: jest.fn()
  };

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly', () => {
    const wrapper = shallow(<Tabs {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should select first item on create when items exist', () => {
    const wrapper = shallow(<Tabs {...props} />);

    const {
      state: { selectedItem }
    } = wrapper.instance();

    const [item] = props.items;

    expect(selectedItem).toBe(item);
  });

  it('should not select first item on create when items do not exist', () => {
    const wrapper = shallow(
      <Tabs
        {...props}
        items={[]}
      />
    );

    const {
      state: { selectedItem }
    } = wrapper.instance();

    expect(selectedItem).toBe(null);
  });

  it('should select item on item click', () => {
    const wrapper = shallow(<Tabs {...props} />);

    const [_, item] = props.items;

    wrapper.find(`.${itemClassName}`).at(1).simulate('click');

    const {
      state: { selectedItem }
    } = wrapper.instance();

    expect(selectedItem).toBe(item);
  });

  it('should call onItemSelect on item click', () => {
    const wrapper = shallow(<Tabs {...props} />);

    const [_, item] = props.items;

    wrapper.find(`.${itemClassName}`).at(1).simulate('click');

    expect(props.onItemSelect).toHaveBeenCalledWith(item);
  });

  it('should render correctly after item selection', () => {
    const wrapper = shallow(<Tabs {...props} />);

    wrapper.find(`.${itemClassName}`).at(1).simulate('click');

    expect(wrapper).toMatchSnapshot();
  });

  describe('selectItem', () => {
    it('should select item', () => {
      const wrapper = shallow(<Tabs {...props} />);

      const [item] = props.items;

      wrapper.instance().selectItem(item)();

      const {
        state: { selectedItem }
      } = wrapper.instance();

      expect(selectedItem).toBe(item);
    });

    it('should call onItemSelect', () => {
      const wrapper = shallow(<Tabs {...props} />);

      const [item] = props.items;

      wrapper.instance().selectItem(item)();

      expect(props.onItemSelect).toHaveBeenCalledWith(item);
    });
  });

  describe('isItemSelected', () => {
    it('should return false when selected item does not exist', () => {
      const wrapper = shallow(<Tabs {...props} />);

      const [item] = props.items;

      wrapper.setState({ selectedItem: null });

      const isItemSelected = wrapper.instance().isItemSelected(item.id);

      expect(isItemSelected).toBeFalsy();
    });

    it('should return true', () => {
      const wrapper = shallow(<Tabs {...props} />);

      const [item] = props.items;

      wrapper.setState({ selectedItem: item });

      const isItemSelected = wrapper.instance().isItemSelected(item.id);

      expect(isItemSelected).toBeTruthy();
    });

    it('should return false', () => {
      const wrapper = shallow(<Tabs {...props} />);

      const [firstItem, secondItem] = props.items;

      wrapper.setState({ selectedItem: firstItem });

      const isItemSelected = wrapper.instance().isItemSelected(secondItem.id);

      expect(isItemSelected).toBeFalsy();
    });
  });

  describe('getItemClassName', () => {
    it('should return item class name', () => {
      const wrapper = shallow(<Tabs {...props} />);

      const [item] = props.items;

      wrapper.instance().isItemSelected = jest.fn().mockReturnValue(false);

      const className = wrapper.instance().getItemClassName(item.id);

      expect(className).toBe(itemClassName);
    });

    it('should return selected item class name', () => {
      const wrapper = shallow(<Tabs {...props} />);

      const [item] = props.items;

      wrapper.instance().isItemSelected = jest.fn().mockReturnValue(true);

      const className = wrapper.instance().getItemClassName(item.id);

      expect(className).toBe(selectedItemClassName);
    });
  });
});
