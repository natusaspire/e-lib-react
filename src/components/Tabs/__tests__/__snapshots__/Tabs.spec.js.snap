// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> should render correctly 1`] = `
<div
  className="tabs"
>
  <div
    className="tabs__item tabs__item--active"
    key="1"
    onClick={[Function]}
  >
    Text
  </div>
  <div
    className="tabs__item"
    key="2"
    onClick={[Function]}
  >
    Audio
  </div>
  <div
    className="tabs__item"
    key="3"
    onClick={[Function]}
  >
    Video
  </div>
</div>
`;

exports[`<Tabs /> should render correctly after item selection 1`] = `
<div
  className="tabs"
>
  <div
    className="tabs__item"
    key="1"
    onClick={[Function]}
  >
    Text
  </div>
  <div
    className="tabs__item tabs__item--active"
    key="2"
    onClick={[Function]}
  >
    Audio
  </div>
  <div
    className="tabs__item"
    key="3"
    onClick={[Function]}
  >
    Video
  </div>
</div>
`;
